### ISSUE TITLE
PyGithub should provide means to throttle API requests

### ISSUE THREAD

**Issue by EnricoMi** [PyGithub/PyGithub#1989]
Sending requests to the GitHub API too quickly will result in abuse rate limit errors. This has been discussed in several issues here (#1235, #2113, #2005, #1319, #1115, #810). I'd like to propose to wrap the Requester.__requestRaw method in a throttling method that ensures subsequent calls are at least a configured number of (float) seconds apart. This way the user of the client library can rely on calls not being too frequent.

Throttling in user code adds complexity on top of the client library and duplicates code across projects that use this client library. A good example is https://github.com/clockfort/GitHub-Backup/pull/44. Especially when iterating over paginated results, throttling becomes difficult in user code.

I have wrapped the `Requester.__requestRaw` method with the following code, which allows me to configure seconds between read (`GET`) and write (all verbs other than `GET`) requests. All requests will then be deferred by the client library if called  too quickly:

https://github.com/EnricoMi/publish-unit-test-result-action/blob/4b6277f1917325e96cadc94a1116031a0ae906bb/python/publish_unit_test_results.py#L99-L122

I am delaying retrys by a urllib `Retry` instance with a backoff factor identical to the configured delays above, which is out of scope of this issue.

**Comment by EnricoMi** [comment]
@s-t-e-v-e-n-k what are your thoughts on this?

**Comment by Tasty213** [comment]
Can I just cast my vote in favour of this? I've got a program that calls many different methods of the library and am currently having to try and wrap all of them in a helper function to prevent hitting the rate limit. Having a simple setting to pass to tell the library just to go slow would both help it conform to the docs which request that you [wait 1 second between requests](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits) when doing a large number of requests.

Could also have an option to wait for the value in the retry-after header instead of having to implement a wait for that project side.

(Btw the library is really great though and really helping with the project I'm working on thanks for the amazing work)

**Comment by C-alcium** [comment]
The way I got around this was to just throw everything into a thread and then cache the results. I think something that doesn't particularly help is it's not 100% clear what results in a request being executed.

**Comment by jogo** [comment]
I was able to use a custom Retry class to handle the retry headers and make sure to trigger a retry on a 403

https://docs.github.com/en/rest/overview/resources-in-the-rest-api#checking-your-rate-limit-status

```python
class RateLimitRetry(urllib3.util.retry.Retry):
    def get_retry_after(self, response):
        reset_time = datetime.datetime.fromtimestamp(int(response.headers["X-RateLimit-Reset"]))
        retry_after = (reset_time - datetime.datetime.now()).seconds + 1
        print(f"Rate limited,  retry after: {retry_after} seconds")
        return retry_after
```

**Comment by LucBerge** [comment]
@jogo How do you use it with PyGithub?

**Comment by EnricoMi** [comment]
Method `get_retry_after` is a nicer place to inject this logic than `increment`. Let me see if I can simplify #2387 with this.

**Comment by EnricoMi** [comment]
@jogo not all 403 responses are rate limit errors, contain a `X-RateLimit-Reset`, or are retryable at all. I think only the content of the response provides enough evidence to identify a retryable 403: #2387

**Comment by jogo** [comment]
@LucBerge I used it with PyGithub the way @EnricoMi has it in his PR. via the `retry` kwarg to `Github`

@EnricoMi thanks for converting this into a PR. 

Based on the [documentation](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#exceeding-the-rate-limit ) I could find checking the header `x-ratelimit-remaining` for a value of 0 should hopefully be enough to determine if it's a rate limit 403 or not.

**Comment by JosephTLyons** [comment]
Hey @EnricoMi, just wanted to thank you for all your work in this area. Over the last 18 months, I kept running into [these errors](https://github.com/zed-industries/zed/actions/runs/8584680293) off and on in my [script for building top-ranking issue reports](https://github.com/zed-industries/zed/blob/main/script/update_top_ranking_issues/main.py) in the Zed editor repository and today I finally decided to try fixing it. I spent a bit of time putting in some custom error handling, but then ran into your PRs here and found things were working after simply bumping the packages (I was on v1.55!). Thanks for all of the work here!
